cli/lsp/urls.rs

```diff
@@ -0,0 +1,186 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+
+use crate::file_fetcher::map_content_type;
+use crate::media_type::MediaType;
+
+use deno_core::error::AnyError;
+use deno_core::url::Url;
+use deno_core::ModuleSpecifier;
+use std::collections::HashMap;
+use std::ops::Deref;
+use std::sync::Arc;
+
+/// This is a partial guess as how URLs get encoded from the LSP.  We want to
+/// encode URLs sent to the LSP in the same way that they would be encoded back
+/// so that we have the same return encoding.
+const LSP_ENCODING: &percent_encoding::AsciiSet = &percent_encoding::CONTROLS
+  .add(b':')
+  .add(b';')
+  .add(b'=')
+  .add(b'@')
+  .add(b'[')
+  .add(b'\\')
+  .add(b']')
+  .add(b'^')
+  .add(b'|');
+
+fn hash_data_specifier(specifier: &ModuleSpecifier) -> String {
+  let url = specifier.as_url();
+  let mut file_name_str = url.path().to_string();
+  if let Some(query) = url.query() {
+    file_name_str.push('?');
+    file_name_str.push_str(query);
+  }
+  crate::checksum::gen(&[file_name_str.as_bytes()])
+}
+
+fn data_url_media_type(specifier: &ModuleSpecifier) -> MediaType {
+  let url = specifier.as_url();
+  let path = url.path();
+  let mut parts = path.splitn(2, ',');
+  let media_type_part =
+    percent_encoding::percent_decode_str(parts.next().unwrap())
+      .decode_utf8_lossy();
+  let (media_type, _) =
+    map_content_type(specifier, Some(media_type_part.into()));
+  media_type
+}
+
+/// A bi-directional map of URLs sent to the LSP client and internal module
+/// specifiers.  We need to map internal specifiers into `deno:` schema URLs
+/// to allow the Deno language server to manage these as virtual documents.
+#[derive(Debug, Default)]
+pub struct LspUrlMap {
+  specifier_to_url: HashMap<Arc<ModuleSpecifier>, Arc<Url>>,
+  url_to_specifier: HashMap<Arc<Url>, Arc<ModuleSpecifier>>,
+}
+
+impl LspUrlMap {
+  pub fn new() -> Self {
+    Default::default()
+  }
+
+  fn put(&mut self, specifier: ModuleSpecifier, url: Url) {
+    let specifier = Arc::new(specifier);
+    let url = Arc::new(url);
+    self.specifier_to_url.insert(specifier.clone(), url.clone());
+    self.url_to_specifier.insert(url, specifier);
+  }
+
+  fn get_url(&self, specifier: &ModuleSpecifier) -> Option<&Url> {
+    self.specifier_to_url.get(specifier).map(Deref::deref)
```

Comment:
```suggestion
    self.specifier_to_url.get(specifier).as_deref()
```
(That's what you're doing here in long-hand, right? Ditto line 75.)

Reply:
That is what I get for copying some old code off stackoverflow.

Reply:
oooh:

```
expected enum `std::option::Option<&deno_core::url::Url>`
found enum `std::option::Option<&Arc<deno_core::url::Url>>`
```

So you gotta unwrap it/map it to get rid of the ref like that.  So it isn't as crazy as it seems.